srtcore/epoll.cpp

```diff
@@ -395,28 +369,25 @@ int CEPoll::wait(const int eid, set<SRTSOCKET>* readfds, set<SRTSOCKET>* writefd
          throw CUDTException(MJ_NOTSUP, MN_EIDINVAL);
       }
 
-      if (p->second.m_sUDTSocksIn.empty() && p->second.m_sUDTSocksOut.empty() && p->second.m_sLocals.empty() && (msTimeOut < 0))
+      if (p->second.m_sUDTSocksWait.empty() && p->second.m_sLocals.empty() && (msTimeOut < 0))
       {
          // no socket is being monitored, this may be a deadlock
          CGuard::leaveCS(m_EPollLock);
          throw CUDTException(MJ_NOTSUP, MN_INVAL);
       }
 
       // Sockets with exceptions are returned to both read and write sets.
-      if ((NULL != readfds) && (!p->second.m_sUDTReads.empty() || !p->second.m_sUDTExcepts.empty()))
-      {
-         *readfds = p->second.m_sUDTReads;
-         for (set<SRTSOCKET>::const_iterator i = p->second.m_sUDTExcepts.begin(); i != p->second.m_sUDTExcepts.end(); ++ i)
-            readfds->insert(*i);
-         total += p->second.m_sUDTReads.size() + p->second.m_sUDTExcepts.size();
-      }
-      if ((NULL != writefds) && (!p->second.m_sUDTWrites.empty() || !p->second.m_sUDTExcepts.empty()))
-      {
-         *writefds = p->second.m_sUDTWrites;
-         for (set<SRTSOCKET>::const_iterator i = p->second.m_sUDTExcepts.begin(); i != p->second.m_sUDTExcepts.end(); ++ i)
-            writefds->insert(*i);
-         total += p->second.m_sUDTWrites.size() + p->second.m_sUDTExcepts.size();
+		for(map<SRTSOCKET, int>::const_iterator it = p->second.m_sUDTSocksSet.begin(); it != p->second.m_sUDTSocksSet.end(); ++ it) {
+			if(readfds && ((it->second & UDT_EPOLL_IN) || (it->second & UDT_EPOLL_ERR))) {
+				if(readfds->insert(it->first).second)
+					++total;
       }
+			if(writefds && ((it->second & UDT_EPOLL_OUT) || (it->second & UDT_EPOLL_ERR))) {
+				if(writefds->insert(it->first).second)
+					++total;
+			}
+		}
+			
```

Comment:
Please check #562. Might be worth fixing it together with these changes, unless it already does.

Reply:
Yes I think it does resolve it.